### PR TITLE
include input fields as output for Substrait consumer

### LINKED ISSUE #12204: Projects from Substrait do not include input fields as output fields
### Describe the bug

According to the [Substrait specification](https://substrait.io/relations/logical_relations/#project-operation) project relations emit all if the input fields followed by the list of new expressions.  Datafusion only emits the new expressions.

### To Reproduce

Pass a Substrait plan such as the following to Datafusion.  (A literal can be used instead of a window function but this is what I had handy.)

```
{
  "extensionUris": [
    {
      "extensionUriAnchor": 1,
      "uri": "/functions_arithmetic.yaml"
    }
  ],
  "extensions": [
    {
      "extensionFunction": {
        "extensionUriReference": 1,
        "functionAnchor": 1,
        "name": "row_number"
      }
    }
  ],
  "relations": [
    {
      "root": {
        "input": {
          "project": {
            "common": {
              "direct": {}
            },
            "input": {
              "read": {
                "common": {
                  "direct": {}
                },
                "baseSchema": {
                  "names": [
                    "user_id",
                    "name",
                    "paid_for_service"
                  ],
                  "struct": {
                    "types": [
                      {
                        "string": {
                          "nullability": "NULLABILITY_REQUIRED"
                        }
                      },
                      {
                        "string": {
                          "nullability": "NULLABILITY_REQUIRED"
                        }
                      },
                      {
                        "bool": {
                          "nullability": "NULLABILITY_REQUIRED"
                        }
                      }
                    ],
                    "nullability": "NULLABILITY_REQUIRED"
                  }
                },
                "namedTable": {
                  "names": [
                    "users"
                  ]
                }
              }
            },
            "expressions": [
              {
                "windowFunction": {
                  "functionReference": 1,
                  "sorts": [
                    {
                      "expr": {
                        "selection": {
                          "directReference": {
                            "structField": {
                              "field": 1
                            }
                          },
                          "rootReference": {}
                        }
                      },
                      "direction": "SORT_DIRECTION_ASC_NULLS_FIRST"
                    }
                  ],
                  "upperBound": {
                    "unbounded": {}
                  },
                  "lowerBound": {
                    "unbounded": {}
                  },
                  "outputType": {
                    "i64": {
                      "nullability": "NULLABILITY_REQUIRED"
                    }
                  },
                  "invocation": 3
                }
              }
            ]
          }
        },
        "names": [
          "user_id",
          "name",
          "paid_for_service",
          "row_number"
        ]
      }
    }
  ],
  "version": {
    "minorNumber": 52,
    "producer": "spark-substrait-gateway"
  }
}
```

### Expected behavior

The result of the plan above would be 4 columns to match the 4 names provided.  The current behavior is that Datafusion returns just one column (row_number) for the project.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Lordworms**
## Which issue does this PR close?

Closes #12204

### CODE REVIEW

**Comment by alamb** [comment]
FYI @Blizzara

**Comment by Blizzara** [comment]
I don't feel this is right - unless I misunderstood something. This tries to fix the problem at the root level of the plan, so that we'd produce the columns the plan asked for, and it works in this test case where there is only a single project in the plan, but I don't think it would work for any more complicated plan. 

I think the proper fix would be around https://github.com/apache/datafusion/blob/53de592381b8585f7d9de82f2ac1eb58b1524e44/datafusion/substrait/src/logical_plan/consumer.rs#L414, to add all of the output fields of the project's input rel into the `exprs` list (and also `names` I guess) before adding the new expressions. Or something along those lines

**Comment by Lordworms** [comment]
> I don't feel this is right - unless I misunderstood something. This tries to fix the problem at the root level of the plan, so that we'd produce the columns the plan asked for, and it works in this test case where there is only a single project in the plan, but I don't think it would work for any more complicated plan.
> 
> I think the proper fix would be around
> 
> https://github.com/apache/datafusion/blob/53de592381b8585f7d9de82f2ac1eb58b1524e44/datafusion/substrait/src/logical_plan/consumer.rs#L414
> 
> , to add all of the output fields of the project's input rel into the `exprs` list (and also `names` I guess) before adding the new expressions. Or something along those lines

Thanks for the suggrestions! I'll take try and adding more test cases

**Comment by vbarua** [comment]
substrait-java has a similar issue with mapping Substrait Projects, which include the input fields in the output, with Calcite Projects which do not. As @Blizzara suggests, a good fix for this is to include all of the input fields as expression in the Project. The following substrait-java code is what handles this https://github.com/substrait-io/substrait-java/blob/c8c31ecd3da047a4de0c33b21f3e960df628bda0/isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java#L155-L171

**Comment by Lordworms** [comment]
> substrait-java has a similar issue with mapping Substrait Projects, which include the input fields in the output, with Calcite Projects which do not. As @Blizzara suggests, a good fix for this is to include all of the input fields as expression in the Project. The following substrait-java code is what handles this https://github.com/substrait-io/substrait-java/blob/c8c31ecd3da047a4de0c33b21f3e960df628bda0/isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java#L155-L171

I think that's what I did in the updated code, just some other points I mentioned in the review

File: datafusion/substrait/src/logical_plan/consumer.rs [hunk 1]
```diff
@@ -214,6 +214,9 @@ pub async fn from_substrait_plan(
                         Ok(from_substrait_rel(ctx, rel, &extensions).await?)
                     },
                     plan_rel::RelType::Root(root) => {
+                        if !root.names.is_empty() {
+                            extensions = extensions.with_projection_names(root.names.clone());
+                        }
```
Comment (vbarua) [inline datafusion/substrait/src/logical_plan/consumer.rs:219]: What's the point of storing the root names here? They only need to be applied to the final output, and that is handled below by the call to `make_renamed_schema`.

File: datafusion/substrait/src/logical_plan/consumer.rs [hunk 2]
```diff
@@ -412,6 +428,10 @@ pub async fn from_substrait_rel(
                 );
                 let mut names: HashSet<String> = HashSet::new();
                 let mut exprs: Vec<Expr> = vec![];
+                input.schema().iter().for_each(|(qualifier, field)| {
+                    exprs.push(col(Column::from((qualifier, field))))
+                });
```
Comment (Blizzara) [inline datafusion/substrait/src/logical_plan/consumer.rs:433]: this seems like the right thing - but then to get tests passing again I think we'll also need to add support for the `remap` https://substrait.io/relations/basics/#emit-output-ordering, both here and in producer...

**Comment by alamb** [comment]
I am trying to clean up the review queue, so marking this PR as draft as the CI tests are failing. Let me know if that wasn't right

**Comment by vbarua** [comment]
I have some work in progress around remaps for https://github.com/apache/datafusion/issues/12347 which I suspect will overlap with this. @Lordworms I'd be happy to pull your changes into my work if that makes sense to you.

**Comment by Lordworms** [comment]
> I have some work in progress around remaps for #12347 which I suspect will overlap with this. @Lordworms I'd be happy to pull your changes into my work if that makes sense to you.

Sure, that would be fine, Thanks a lot

**Comment by vbarua** [comment]
I believe this can be closed following https://github.com/apache/datafusion/pull/13127

**Comment by alamb** [comment]
Thanks @vbarua  -- @Lordworms  please let us know if this isn't right to close

**Pull request closed without merging**